### ISSUE TITLE
Micromamba dry-run JSON occasionally omits hash values

### ISSUE THREAD

**Issue by maresb** [mamba-org/mamba#2346]
### Troubleshooting docs

- [X] My problem is not solved in the Troubleshooting docs

### Search tried in issue tracker

n/a

### Latest version of Mamba

- [X] My problem is not solved with the latest version

### Tried in Conda?

Not reproducible with Conda

### Describe your issue

This is a followup from https://github.com/conda/conda-lock/pull/376 where we have been regularly encountering sporadic failures from Micromamba when our test suite runs on GitHub Actions.

Our test suite runs parallelized, spawning several Conda/Mamba/Micromamba instances, but I am unsure whether or not this issue is due to interprocess interference.

This feels like it may be some sort of race condition where metadata is not completely available due to a package still being in the process of extraction.

An example instance of this problem occurs in [this run](https://github.com/conda/conda-lock/actions/runs/4275571616/jobs/7443025517) on MacOS AMD running conda-lock with Python 3.11. There is nothing special about this configuration, except that it arises on Linux only very rarely.

One interesting observation is that when this error occurs, I see the following repeated 62 times in Micromamba's `stderr`:
```
mapping pread: No such file or directory
```

The command we run is
```python
['/Users/runner/Library/Application Support/ensure-conda/micromamba', 'create', '--prefix', '/Users/runner/work/_temp/tmptb6fjy74/prefix', '--dry-run', '--json', '-v', '--override-channels', '--channel', 'conda-forge', '--channel', 'file:///Users/runner/work/_temp/tmpb5nn0fjz', 'python 3.9.*', 'dask-core', 'pip *']
```
where the essential bit seems to be
```bash
micromamba create --prefix=/tmp/asdf --dry-run --json -c conda-forge some-library
```

We parse the JSON output of this command, which primarily comprises `actions.LINK` and `actions.FETCH` of various packages. It seems that `actions.FETCH` pertains to the details of packages which have not been cached, but honestly I don't really understand this structure. But to provide a concrete example, if I run

```bash
docker run --rm -it mambaorg/micromamba:1.3.1
micromamba create --dry-run --json -v --channel=conda-forge --name=x libgomp
micromamba install -c conda-forge libgomp
micromamba create --dry-run --json -v --channel=conda-forge --name=x libgomp
```

then the first `micromamba create` command outputs

```json
{
    "actions": {
        "FETCH": [
            {
                "build": "conda_forge",
                "build_number": 0,
                "build_string": "conda_forge",
                "channel": "https://conda.anaconda.org/conda-forge/linux-64",
                "constrains": null,
                "depends": null,
                "fn": "_libgcc_mutex-0.1-conda_forge.tar.bz2",
                "license": "None",
                "md5": "d7c89558ba9fa0495403155b64376d81",
                "name": "_libgcc_mutex",
                "sha256": "fe51de6107f9edc7aa4f786a70f4a883943bc9d39b3bb7307c04c41410990726",
                "size": 2562,
                "subdir": "linux-64",
                "timestamp": 1578324546,
                "track_features": "",
                "url": "https://conda.anaconda.org/conda-forge/linux-64/_libgcc_mutex-0.1-conda_forge.tar.bz2",
                "version": "0.1"
            },
            {
                "build": "h65d4601_19",
                "build_number": 19,
                "build_string": "h65d4601_19",
                "channel": "https://conda.anaconda.org/conda-forge/linux-64",
                "constrains": null,
                "depends": [
                    "_libgcc_mutex 0.1 conda_forge"
                ],
                "fn": "libgomp-12.2.0-h65d4601_19.tar.bz2",
                "license": "GPL-3.0-only WITH GCC-exception-3.1",
                "md5": "cedcee7c064c01c403f962c9e8d3c373",
                "name": "libgomp",
                "sha256": "81a76d20cfdee9fe0728b93ef057ba93494fd1450d42bc3717af4e468235661e",
                "size": 466188,
                "subdir": "linux-64",
                "timestamp": 1666519598,
                "track_features": "",
                "url": "https://conda.anaconda.org/conda-forge/linux-64/libgomp-12.2.0-h65d4601_19.tar.bz2",
                "version": "12.2.0"
            }
        ],
        "LINK": [
            {
                "build": "conda_forge",
                "build_number": 0,
                "build_string": "conda_forge",
                "channel": "https://conda.anaconda.org/conda-forge/linux-64",
                "constrains": null,
                "depends": null,
                "fn": "_libgcc_mutex-0.1-conda_forge.tar.bz2",
                "license": "None",
                "md5": "d7c89558ba9fa0495403155b64376d81",
                "name": "_libgcc_mutex",
                "sha256": "fe51de6107f9edc7aa4f786a70f4a883943bc9d39b3bb7307c04c41410990726",
                "size": 2562,
                "subdir": "linux-64",
                "timestamp": 1578324546,
                "track_features": "",
                "url": "https://conda.anaconda.org/conda-forge/linux-64/_libgcc_mutex-0.1-conda_forge.tar.bz2",
                "version": "0.1"
            },
            {
                "build": "h65d4601_19",
                "build_number": 19,
                "build_string": "h65d4601_19",
                "channel": "https://conda.anaconda.org/conda-forge/linux-64",
                "constrains": null,
                "depends": [
                    "_libgcc_mutex 0.1 conda_forge"
                ],
                "fn": "libgomp-12.2.0-h65d4601_19.tar.bz2",
                "license": "GPL-3.0-only WITH GCC-exception-3.1",
                "md5": "cedcee7c064c01c403f962c9e8d3c373",
                "name": "libgomp",
                "sha256": "81a76d20cfdee9fe0728b93ef057ba93494fd1450d42bc3717af4e468235661e",
                "size": 466188,
                "subdir": "linux-64",
                "timestamp": 1666519598,
                "track_features": "",
                "url": "https://conda.anaconda.org/conda-forge/linux-64/libgomp-12.2.0-h65d4601_19.tar.bz2",
                "version": "12.2.0"
            }
        ],
        "PREFIX": "/opt/conda/envs/x"
    },
    "dry_run": true,
    "prefix": "/opt/conda/envs/x",
    "success": true
}
```

where `actions.LINK == actions.FETCH`. In the second 
while the second `micromamba create` has the exact same output except that `actions.FETCH` has been deleted.

For the particular run in question, the following are the fetch action names:
```python
{
  'python', 'ca-certificates', 'libgomp', 'libgcc-ng', 'toolz', 'libnsl', 'ld_impl_linux-64', 
  'ncurses', 'bzip2', 'yaml', 'partd', 'click', 'openssl', 'pyyaml', 'tk', 'libffi', 
  '_openmp_mutex', '_libgcc_mutex', 'cloudpickle', 'libuuid', 'dask-core', 'libsqlite', 
  'readline', 'libzlib', 'python_abi', 'locket', 'fsspec', 'xz'
}
```
and the link actions are the fetch actions plus the following "link-only actions": 
```python
{
  'tzdata', 'pip', 'wheel', 'setuptools', 'packaging'
}
```

The actions where `md5` and `sha256` are missing are the link-only actions plus the following fetch actions:
```python
{
  'locket', 'fsspec', 'cloudpickle', 'toolz', 'click', 'partd', 'dask-core'
}
```

The link actions with `md5` and `sha256` missing look like

```json
  {
    "build": "h191b570_0",
    "build_number": 0,
    "build_string": "h191b570_0",
    "channel": "https://conda.anaconda.org/conda-forge/noarch",
    "constrains": null,
    "depends": null,
    "fn": "tzdata-2022g-h191b570_0.conda",
    "license": "LicenseRef-Public-Domain",
    "name": "tzdata",
    "size": 108083,
    "subdir": "noarch",
    "timestamp": 1669765202,
    "track_features": "",
    "url": "https://conda.anaconda.org/conda-forge/noarch/tzdata-2022g-h191b570_0.conda",
    "version": "2022g"
  }
```

so the expected keys minus `md5` and `sha256`. Note that both `.conda` and `.tar.bz2` archives show this phenomenon.

I strongly suspect that this has to do with a similar issue we also see with Conda, for example in [this run](https://github.com/conda/conda-lock/actions/runs/4274895569/jobs/7441851465). In this case, we are parsing the output of `conda create --dry-run --json ...`. When using Conda as opposed to Micromamba, the link records contain less information than the fetch records. Thus in this case we try to read the data out of the `repodata_record.json` file. Sometimes the file cannot be found. In this particular run, we hit the issue with `libgomp`. The following related files can be found in the pkgs_dir:

```
/home/runner/conda_pkgs_dir/libgomp-12.2.0-h65d4601_19.tar.bz2
/home/runner/conda_pkgs_dir/libgomp-12.2.0-h65d4601_19
/home/runner/conda_pkgs_dir/libgomp-12.2.0-h65d4601_19.tar.bz2.c~
```

If I wait 100ms, then the `repodata_record.json` file appears.

My suspicion is that the Micromamba failure occurs when Micromamba internally encounters a similar situation, where perhaps some metadata goes missing because some external `conda install` process has not entirely completed. But I'm not entirely sure.

Any ideas what's going on? Any desire to pursue this further?

The workaround I implemented on the conda-lock side is to source the metadata from `repodata_record.json` rather than Micromamba's output, and if the `repodata_record.json` does not exist, then retry after 100 ms for up to 1 sec. It's not particularly elegant, but at least our CI seems stable now.

**Comment by jaimergp** [comment]
The `FETCH` phase is reading from repodata.json, the `LINK` phase reads from the in-archive `repodata_record.json`. The in-archive version does not know the sha256/md5 because it's created _before_ compressing the file. It's a design limitation, I'd say, but I agree that it would be better to push the repodata.json entry (to also account for repodata patches!). This happens in `conda` too, AFAICR.

**Comment by maresb** [comment]
Thanks so much @jaimergp for chiming in here. I still don't get it though:

> The in-archive version does not know the sha256/md5 because it's created before compressing the file.

What's being compressed? Aren't we dealing with a downloaded artifact that has hashes determined at build-time and recorded in `repodata.json`? Are they being discarded after the download because we only do the verification after download?

**Comment by jaimergp** [comment]
Sorry, I typed my message from mobile so it was a bit terse. Let me try again with some links:


- The `LINK` section is populated with the `repodata_record.json` present in each extracted archive, inside `info/`. We have these files locally because they are in the `pkgs/` cache. It does not know about the hashes because these JSONs were created by conda-build _before_ the archive was packed and compressed, so there's no way to know.
- The `FETCH` section is populated with the packages not present in the `pkgs/` cache. As such, the only source for the PackageRecord data is the entry in the `repodata.json`. This does have the hashes because they are added by conda-index when the archives (.conda, .tar.bz2) are processed. conda-index extracts the inner `repodata_record.json` from `info/` and extends it with the calculated hashes of the compressed artifacts. `repodata.json` is created. Something like:


```python
for file in os.listdir(channel):
	if not file.endswith((".tar.bz2", ".conda")):
       continue
	record_data = json.loads(extract_archive(file) / "info" / "repodata_record.json")
    record_data["sha256"] = compute_hash(file, "sha256")
    record_data["md5sum"] = compute_hash(file, "md5sum")
    yield record_data
```

I assume there's a convenience / laziness reason to not map the `info/repodata_record.json` to the equivalent entry in `repodata.json`. Since the `pkgs/` cache is not aware of channels (it's just a flat collection of archives and their extracted diretories), maybe there's no (guaranteed) way to map the files back to channel repodata.json?

**Comment by jaimergp** [comment]
> Actually, channel somehow seems to be included in about.json. 

I think these are the channels that were used to source the packages installed in the build/host prefixes, not the target one. In the case of conda-forge, we do see the target channels, but that's an implementation detail of how conda-forge's infra, and not guaranteed to be in every conda package:

<img width="509" alt="image" src="https://github.com/user-attachments/assets/dc94f6b8-1a84-4b6b-bc84-f37b09d61fda" />

**Comment by maresb** [comment]
So if I understand correctly, it would in principle be possible to carry over the metadata from repodata.json, but there's simply no obvious place for it. So you'd need something like sidecar files :grimacing:

**Comment by jaimergp** [comment]
It would be possible and actually desired, but there might be weird edge cases until we figure out a channel-aware cache, but that will have ripple effects in what we consider the "identity" of a package; e.g. if the same package is copied from one channel to another, should it have one or two cache entries? Maybe we should just make the cache indexed by sha256 instead of their filename 🤷 That's a topic for a CEP though.

For now, we (speaking from `conda/conda`, which also suffers from this issue), we could try to query the repodata with the given filename and hope for the best. BUT that begs the question whether we can just "break" the `--json` schema by adding some fields in `LINK` (probably ok) and / or (perhaps more controversial) have it display the patched repodata if that applies.
